pkg/agent/nodeportlocal/k8s/controller.go

```diff
@@ -0,0 +1,266 @@
+
+// Copyright 2020 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8s
+
+import (
+	"encoding/json"
+	"fmt"
+	"time"
+
+	"github.com/vmware-tanzu/antrea/pkg/agent/nodeportlocal/portcache"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+)
+
+const (
+	minRetryDelay = 2 * time.Second
+	maxRetryDelay = 120 * time.Second
+	numWorkers    = 4
+)
+
+type Controller struct {
+	portTable  *portcache.PortTable
+	kubeClient clientset.Interface
+	queue      workqueue.RateLimitingInterface
+	Ctrl       cache.Controller
+	CacheStore cache.Store
+	PodToIP    map[string]string
+}
+
+func NewNPLController(kubeClient clientset.Interface, pt *portcache.PortTable) *Controller {
+	nplCtrl := Controller{
+		kubeClient: kubeClient,
+		portTable:  pt,
+		PodToIP:    make(map[string]string),
+	}
+	nplCtrl.queue = workqueue.NewNamedRateLimitingQueue(workqueue.NewItemExponentialFailureRateLimiter(minRetryDelay, maxRetryDelay), "nodeportlocal")
+	return &nplCtrl
+}
+
+func podKeyFunc(pod *corev1.Pod) string {
+	return pod.Namespace + "/" + pod.Name
+}
+
+// Run starts to watch and process Pod updates for the Node where Antrea Agent is running.
+// It starts a queue and a fixed number of workers to process the objects from the queue.
+func (c *Controller) Run(stopCh <-chan struct{}) {
+	defer c.queue.ShutDown()
+	for i := 0; i < numWorkers; i++ {
+		go wait.Until(c.Worker, time.Second, stopCh)
+	}
+	c.Ctrl.Run(stopCh)
+	cache.WaitForCacheSync(stopCh, c.Ctrl.HasSynced)
+	<-stopCh
+}
+
+func (c *Controller) syncPod(key string) error {
+	obj, exists, err := c.CacheStore.GetByKey(key)
+	if err != nil {
+		return err
+	} else if exists {
+		return c.HandleAddUpdatePod(key, obj)
+	} else {
+		return c.HandleDeletePod(key)
+	}
+}
+
+func (c *Controller) checkDeletedPod(obj interface{}) (*corev1.Pod, error) {
+	deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+	if !ok {
+		return nil, fmt.Errorf("Received unexpected object: %v", obj)
+
+	}
+	pod, ok := deletedState.Obj.(*corev1.Pod)
+	if !ok {
+		return nil, fmt.Errorf("DeletedFinalStateUnknown object is not of type Pod: %v", deletedState.Obj)
+	}
+	return pod, nil
+}
+
+func (c *Controller) EnqueueObjAdd(obj interface{}) {
+	pod, isPod := obj.(*corev1.Pod)
+	if !isPod {
+		var err error
+		pod, err = c.checkDeletedPod(obj)
```

Comment:
I think it's impossible to get "DeletedFinalStateUnknown" from add event.
And if this is to reuse code with `EnqueueObjDel`, we should just have single `EnqueueObj` for the 3 handlers like:
```
		cache.ResourceEventHandlerFuncs{
			AddFunc:    c.EnqueueObj,
			DeleteFunc: c.EnqueueObj,
			UpdateFunc: func(old, cur interface{}) {c.EnqueueObj(cur)},
		},
```